content/blog/2020-12-23-hof/index.md

```diff
@@ -0,0 +1,83 @@
+
+`apply` : Indicates a function application. The first argument is the function name. The subsequent arguments are arguments to that function.
+
+Here's an example of those two operations being used in a Bril program: 
+
+```
+a : int = const 1; 
+b : int = const 2;
+add : fun<int, int>, <int> = anon(x, y) {
+        res : int = x + y; 
+        ret res;
+    };
+res : int = apply add a b;
+```
+
+### Implementation 
+
+#### Intuition 
+
+Lambda lifting as an idea is pretty straightforward. First, make sure all the functions are uniquely named. Then, do the following until there are no more free variables or local functions: Replace all free variables to the higher-order function with an extra argument to the outer function that gets passed in to the higher-order one. Then, pull out every local function definition without free variables into the global scope.
+
+#### Code Breakdown
+
+I had three main functions: `free_vars`, which took in a function's arguments and body and determined the free variables it contained, `process_func`, which took in a function's arguments as well as its basic blocks and returned a list of extracted global functions as well as the original function's new block structure, and `process_q`, which took in a program's CFG and queued each of its functions for processing with `process_func`, adding on new global functions to extract other functions from to the queue. 
+
+#### `bril-txt` modification 
+
+I updated `briltxt.py` to translate Bril programs using HOFs in a textual representation into the JSON representation (and vice versa). This made it a lot easier to write and run test cases! 
+
+## What was hard to get right? 
+
+A saying (by Phil Karlton, allegedly) that gets riffed on a lot among computer scientists goes: *There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors*. I encountered all of the above when implementing HOFs in Bril. 
+
+Function renaming was tricky as the mangling scheme I implemented made it hard to keep track of which function names mapped to each function in my cache of functions. I ended up scrapping the mangling scheme as it was never a problem in the example programs and I figured I could require unique names as part of the specification. I also had some issues with queueing extracted global functions into my lambda lifter; I was always missing the last one. I'm not sure if this was because I was working in OCaml where mutability is not welcomed, but after making the code more explicitly imperative this problem vanished. 
+
+I discovered that I couldn't take functions as arguments or return them since that would create a closure, which is a more general form of the nested functions I was implementing. To implement the desired functionality, I would need to develop a closure syntax, which was beyond what I was able to accomplish this semster. 
```

Comment:
Perhaps more importantly than closure syntax, you'd need environment _values_ somehow. Like, a struct to store the closed-over state. In fact, maybe structs as proposed elsewhere would suffice.

Reply:
I see what you're saying. Let me know if this looks okay now!

Reply:
Yep!